sql: add retries on COPY under certain conditions

otan opened the pull request:
Release note (sql change): If `copy_from_retries_enabled` is set, COPY is now able to retry certain safe circumstances - namely when `copy_from_atomic_enabled` is false, there is no transaction running COPY and the error returned is retriable. This prevents users who keep running into `TransactionProtoWithRefreshError` from having issues.

Informs #90656

Release justification: feature gated new feature critical for unblocking DMS

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95275)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95275)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @otan)

___
*[`pkg/sql/copy.go` line 840 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95275#-NLyZwh42KbbNgbsLLN2:-NLyZwh42KbbNgbsLLN3:b-fgendl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5178f8258b2d2817f7d6e577f7caaa829edaa23/pkg/sql/copy.go#L840)):*
> ```Go
> 			// It is currently only safe to retry if we are not in atomic copy mode &
> 			// we are in an implicit transaction.
> 			if c.implicitTxn && !c.p.SessionData().CopyFromAtomicEnabled && c.p.SessionData().CopyFromRetriesEnabled && errIsRetriable(err) {
> ```

this could retry an infinite number of times right? should we add a cap?

also, isn't there already similar retry logic in the conn executor?i guess that part of the logic isn't shared with COPY?
___
*[`pkg/sql/vars.go` line 2326 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95275#-NLy_NIK6maIf73stOu6:-NLy_NILDaN_dddiFVAv:b-nj4s6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5178f8258b2d2817f7d6e577f7caaa829edaa23/pkg/sql/vars.go#L2326)):*
> ```Go
> 
> 	// CockroachDB extension.
> 	`copy_from_retries_enabled`: {
> ```

should it be `Hidden: true` ?
___
*[`pkg/sql/sessiondatapb/local_only_session_data.proto` line 289 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95275#-NLy_WJCCDoPg7XaXQbC:-NLy_WJCCDoPg7XaXQbD:bqsksc4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5178f8258b2d2817f7d6e577f7caaa829edaa23/pkg/sql/sessiondatapb/local_only_session_data.proto#L289)):*
> ```protobuf
>   // CopyFromRetriesEnabled controls whether retries should be internally
>   // attempted for retriable errors.
>   bool copy_from_retries_enabled = 89;
> ```

nit: do you think it would be easier for the next one here if this was at the end of the struct?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95275)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @rafiss)

___
*[`pkg/sql/copy.go` line 840 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95275#-NLyZwh42KbbNgbsLLN2:-NLzCzwr3fzjYlpYZKz7:bqycxi4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5178f8258b2d2817f7d6e577f7caaa829edaa23/pkg/sql/copy.go#L840)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this could retry an infinite number of times right? should we add a cap?

also, isn't there already similar retry logic in the conn executor?i guess that part of the logic isn't shared with COPY?
</blockquote></details>

Nope - the `retry.StartWithCtx` only goes until the backoff reaches 2s.

we can't re-use that logic because COPY is special and consumes the read buffer, so when the conn executor comes around again the COPY data is already gone.
___
*[`pkg/sql/vars.go` line 2326 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95275#-NLy_NIK6maIf73stOu6:-NLzD9BV3cxyfCzJ6Tau:b-fz7q8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5178f8258b2d2817f7d6e577f7caaa829edaa23/pkg/sql/vars.go#L2326)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should it be `Hidden: true` ?
</blockquote></details>

I prefer it being available and read since we're going to advertise using it.
___
*[`pkg/sql/sessiondatapb/local_only_session_data.proto` line 289 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95275#-NLy_WJCCDoPg7XaXQbC:-NLzDC8U1C5fwweHTtuX:bgp36p5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5178f8258b2d2817f7d6e577f7caaa829edaa23/pkg/sql/sessiondatapb/local_only_session_data.proto#L289)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: do you think it would be easier for the next one here if this was at the end of the struct?
</blockquote></details>

easier to backport this way but sure!


<!-- Sent from Reviewable.io -->

otan commented:
*[`pkg/sql/copy.go` line 840 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95275#-NLyZwh42KbbNgbsLLN2:-NLzDvnB3L6weL_uNT-i:b-nfoz1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5178f8258b2d2817f7d6e577f7caaa829edaa23/pkg/sql/copy.go#L840)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Nope - the `retry.StartWithCtx` only goes until the backoff reaches 2s.

we can't re-use that logic because COPY is special and consumes the read buffer, so when the conn executor comes around again the COPY data is already gone.
</blockquote></details>

added a note to that effect.


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95275)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @otan)

___
*[`pkg/sql/copy.go` line 840 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95275#-NLyZwh42KbbNgbsLLN2:-NM-LsgZ1gceyBBalaM2:b-jd0j4t) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5178f8258b2d2817f7d6e577f7caaa829edaa23/pkg/sql/copy.go#L840)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

added a note to that effect.
</blockquote></details>

i'm reading it differently. i see that `retry.StartWithCtx` only exits the retry loop if:
- the current attempt is >= `MaxRetries`, which is not set in `base.DefaultRetryOptions()`
- or if the retry ctx is cancelled

 the `MaxBackoff` you are talking about only limits the max time in between retry attempts, but the number of retry attempts could still be infinite.


<!-- Sent from Reviewable.io -->

otan commented:
*[`pkg/sql/copy.go` line 840 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95275#-NLyZwh42KbbNgbsLLN2:-NM0Inlb4USefsUOlF6K:b-lxy8d8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5178f8258b2d2817f7d6e577f7caaa829edaa23/pkg/sql/copy.go#L840)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i'm reading it differently. i see that `retry.StartWithCtx` only exits the retry loop if:
- the current attempt is >= `MaxRetries`, which is not set in `base.DefaultRetryOptions()`
- or if the retry ctx is cancelled

 the `MaxBackoff` you are talking about only limits the max time in between retry attempts, but the number of retry attempts could still be infinite.
</blockquote></details>

good catch, added the MaxRetries + added a test


<!-- Sent from Reviewable.io -->

cucaroach reviewed (approved):
Can you update this comment: https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/conn_executor.go#L2555

<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 5 of 8 files at r1, 1 of 2 files at r2, 1 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95275#-:-NM0Mj2h1qG9tQDOLz4F:b-yljmkd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/copy.go` line 840 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95275#-NLyZwh42KbbNgbsLLN2:-NM0LfDh3i5A3U0ZJ65h:bgcdvez) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5178f8258b2d2817f7d6e577f7caaa829edaa23/pkg/sql/copy.go#L840)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

good catch, added the MaxRetries + added a test
</blockquote></details>

In theory we could also retry an atomic copy if c.insertedRows is zero no?  Not saying its a good idea but thought I'd mention it.


<!-- Sent from Reviewable.io -->

otan committed:
sql: add retries on COPY under certain conditions

Release note (sql change): If `copy_from_retries_enabled` is set, COPY
is now able to retry certain safe circumstances - namely when
`copy_from_atomic_enabled` is false, there is no transaction running COPY
and the error returned is retriable. This prevents users who
keep running into `TransactionProtoWithRefreshError` from having issues.

otan commented:
*[`pkg/sql/copy.go` line 840 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95275#-NLyZwh42KbbNgbsLLN2:-NM0O2Q96lqfsCcN8uUb:bv6hty9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5178f8258b2d2817f7d6e577f7caaa829edaa23/pkg/sql/copy.go#L840)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

In theory we could also retry an atomic copy if c.insertedRows is zero no?  Not saying its a good idea but thought I'd mention it.
</blockquote></details>

i'll add a TODO


<!-- Sent from Reviewable.io -->

otan commented:
comment updated!





<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
> Reviewed 5 of 8 files at r1, 1 of 2 files at r2, 1 of 2 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95275)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @otan)



<!-- Sent from Reviewable.io -->

otan commented:
bors r=rafiss,cucaroach
